### PR TITLE
[benchmarking] Time execution

### PR DESCRIPTION
**Pull request by daniel-goldstein**
- Time execution stages in the backend (optimization, compilation, runtime)
- Return a dictionary back to the front end of nanoseconds and formatted times for each stage.
- Add `hl.eval_timed` which propagates returns the evaluated IR as well as the timings dictionary. If called with `timed=False` (default), `Env.backend.execute()` drops the timings it received from the backend. Could change it to log them instead or always.

### CODE REVIEW

File: hail/src/main/scala/is/hail/backend/spark/SparkBackend.scala [hunk 1]
```diff
@@ -5,40 +5,59 @@ import is.hail.annotations.{Region, SafeRow}
...
+import org.json4s.jackson.{JsonMethods, Serialization}
 
 object SparkBackend {
   def executeJSON(ir: IR): String = {
     val t = ir.typ
-    val value = execute(ir)
-    JsonMethods.compact(JSONAnnotationImpex.exportAnnotation(value, t))
+    val (value, timings) = execute(ir)
+    val jsonValue = JsonMethods.compact(JSONAnnotationImpex.exportAnnotation(value, t))
+
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/backend/spark/SparkBackend.scala]: we should write all the timing information to the log here

File: hail/src/main/scala/is/hail/backend/local/LocalBackend.scala [hunk 2]
```diff
@@ -2,37 +2,48 @@ package is.hail.backend.local
...
 
-    ir = Optimize(ir, noisy = true, canGenerateLiterals = false, context = Some("LocalBackend.execute - after TableIR lowering"))
+    ir = timer.time(
+      Optimize(ir, noisy = true, canGenerateLiterals = false, context = Some("LocalBackend.execute - after TableIR lowering")),
+      "optimize after table lowering")
 
     println("LocalBackend.execute", Pretty(ir))
 
-    Interpret[Any](ir)
+    val value = timer.time(Interpret[Any](ir), "runtime")
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/backend/local/LocalBackend.scala]: we should write all the timing information to the log here

Comment (daniel-goldstein) [reply hail/src/main/scala/is/hail/backend/local/LocalBackend.scala]: Is there a reason this is a different location than in the `SparkBackend`? I logged at the top-most function of both

Comment (tpoterba) [reply hail/src/main/scala/is/hail/backend/local/LocalBackend.scala]: oh, yeah, that works

File: hail/src/main/scala/is/hail/backend/local/LocalBackend.scala [hunk 3]
```diff
@@ -2,37 +2,50 @@ package is.hail.backend.local
...
 
 object LocalBackend {
   def executeJSON(ir: IR): String = {
     val t = ir.typ
-    val value = execute(ir)
-    JsonMethods.compact(
-      JSONAnnotationImpex.exportAnnotation(value, t))
+    val (value, timings) = execute(ir)
+    val jsonValue = JsonMethods.compact(JSONAnnotationImpex.exportAnnotation(value, t))
+    log.info(s"Execution times: ${ timings.value }")
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/backend/local/LocalBackend.scala]: this isn't going to be particularly readable -- maybe add a function on `Timings` for logging that prints the readable times with one `log.info` call per key?

Comment (daniel-goldstein) [reply hail/src/main/scala/is/hail/backend/local/LocalBackend.scala]: Good point, printed the readable time, one stage at a time

**Pull request merged**